sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala

```diff
@@ -303,4 +303,35 @@ private[spark] object SchemaUtils {
       case _ => false
     }
   }
+
+  /**
+   * Replaces any collated string type with non collated StringType
+   * recursively in the given schema.
+   */
+  def replaceCollatedStringWithStringInSchema(schema: StructType): StructType = {
+    StructType(schema.map { field =>
+      if (hasNonBinarySortableCollatedString(field.dataType)) {
```

Comment:
I think `binarySortable` is not the property we are looking for. We need to fine collation id != 0, according to how `StringType.typeName` is implemented.

Reply:
i'm not a fan of the name `hasNonBinarySortableCollatedString`, but that method does what you described it will check if a collation `supportsBinaryEquality` => if id == 0

Reply:
i'm not a fan of the name, but that is what the function will actually do. It will check if the collation `supportsBinaryOrdering` which only UTF8_BINARY collation does

Reply:
I don't think this is future-proof. I think it's better to write new code to explicitly check `collationId != 0`

Reply:
when a wrote this method a while back it was called `hasNonDefaultCollatedString` and the logic was `collationId != 0`.  However the name was changed a while back by @mihailom-db when he introduced a concept of default collation (session collation).

All in all, method here should be used for exactly this use-case but we revisit naming/logic of it separately

Reply:
I fixed this in a separate PR so please check it out #46116